core/src/main/java/org/elasticsearch/monitor/fs/FsProbe.java

```diff
@@ -135,16 +135,24 @@ public FsInfo stats(FsInfo previous, @Nullable ClusterInfo clusterInfo) throws I
         return Files.readAllLines(PathUtils.get("/proc/diskstats"));
     }
 
+    /** See: https://bugs.openjdk.java.net/browse/JDK-8162520 */
+    private static long adjustForHugeFilesystems(long bytes) {
+        if (bytes < 0) {
+            return Long.MAX_VALUE;
+        }
+        return bytes;
+    }
+
     public static FsInfo.Path getFSInfo(NodePath nodePath) throws IOException {
         FsInfo.Path fsPath = new FsInfo.Path();
         fsPath.path = nodePath.path.toAbsolutePath().toString();
 
         // NOTE: we use already cached (on node startup) FileStore and spins
         // since recomputing these once per second (default) could be costly,
         // and they should not change:
-        fsPath.total = nodePath.fileStore.getTotalSpace();
-        fsPath.free = nodePath.fileStore.getUnallocatedSpace();
-        fsPath.available = nodePath.fileStore.getUsableSpace();
+        fsPath.total = adjustForHugeFilesystems(nodePath.fileStore.getTotalSpace());
+        fsPath.free = adjustForHugeFilesystems(nodePath.fileStore.getUnallocatedSpace());
+        fsPath.available = adjustForHugeFilesystems(nodePath.fileStore.getUsableSpace());
```

Comment:
So I am on the fence about these two, `free` and `available`, because I have seen filesystems where a negative "free" value is legal (I believe ZFS has returned negative free values before for disk space), so I'm not sure whether we should drop these.

I do think we should have the protection on `getTotalSpace` though, I don't think it's as likely for it to be negative and still a legal value.